server/controllers/paypal.js

```diff
@@ -16,193 +15,121 @@ const {
-      .catch(cb);
-    }],
-
-    createPayload: ['createPaymentMethod', function(cb, results) {
-      const payload = {
-        currencyCode: 'USD',
-        startingDate: new Date().toISOString(),
-        endingDate,
-        returnUrl,
-        cancelUrl,
-        displayMaxTotalAmount: false,
-        feesPayer: 'SENDER',
-        maxTotalAmountOfAllPayments,
-        requestEnvelope: {
-          errorLanguage:  'en_US'
-        },
-        clientDetails: results.createPaymentMethod.id
-      };
-      return cb(null, payload);
-    }],
-
-    callPaypal: ['createPayload', function(cb, results) {
-      paypalAdaptive.preapproval(results.createPayload)
-      .then(response => cb(null, response))
-      .catch(cb)
-    }],
-
-    updatePaymentMethod: ['createPaymentMethod', 'createPayload', 'callPaypal', function(cb, results) {
-      const paymentMethod = results.createPaymentMethod;
-      paymentMethod.token = results.callPaypal.preapprovalKey;
-      paymentMethod.save()
-        .then(paymentMethod => cb(null, paymentMethod))
-        .catch(cb);
-    }]
-
-  }, (err, results) => {
-    if (err) return next(err);
-    res.json(results.callPaypal);
-  });
-
+  const startDate = new Date();
+  const endDate = (req.query.endingDate && (new Date(req.query.endingDate))) || moment().add(1, 'years');
+  
+  const payload = {
+    currencyCode: 'USD', // TODO: figure out if there is a reliable way to specify correct currency for a HOST.
+    startingDate: startDate.toISOString(),
+    endingDate: endDate.toISOString(),
+    returnUrl,
+    cancelUrl,
+    displayMaxTotalAmount: false,
+    feesPayer: 'SENDER',
+    maxAmountPerPayment: 2500.00, // PayPal claims this can go up to $10k without needing additional permissions from them.
+    clientDetails: req.remoteUser.id
+  };
+
+  let response;
+
+  return paypalAdaptive.preapproval(payload)
+  .tap(r => response = r)
+  .then(response => PaymentMethod.create({
+    service: 'paypal',
+    UserId: req.remoteUser.id,
+    token: response.preapprovalKey,
+    startDate,
+    endDate
+  }))
+  .then(() => res.json(response))
+  .catch(next);
 };
 
 /**
  * Confirm a preapproval.
- * TODO: remove async and simplify with promises #promisify
  */
 export const confirmPreapproval = function(req, res, next) {
+  let paymentMethod;
+
+  // fetch original payment method
+  return PaymentMethod.findOne({
+    where: {
+      service: 'paypal',
+      UserId: req.remoteUser.id,
+      token: req.params.preapprovalkey
+    }
+  })
+  .tap(pm => paymentMethod = pm)
+  .then(pm => pm ? 
+    Promise.resolve() : 
+    Promise.reject(new errors.NotFound('This preapprovalKey does not exist.')))
+
+  .then(() => getPreapprovalDetailsAndUpdatePaymentMethod(req.params.preapprovalkey, req.remoteUser.id, paymentMethod))
+
+  .then(() => Activity.create({
+    type: 'user.paymentMethod.created',
+    UserId: req.remoteUser.id,
+    data: {
+      user: req.remoteUser.minimal,
+      paymentMethod
```

Comment:
not sure it's safe to store the paymentMethod in the activity. We should sanitize it first.